server/src/main/java/org/opensearch/common/util/MovingAverage.java

```diff
@@ -17,19 +17,47 @@ public class MovingAverage {
     private final int windowSize;
     private final long[] observations;
 
-    private long count = 0;
-    private long sum = 0;
-    private double average = 0;
+    private volatile long count = 0;
+    private volatile long sum = 0;
+    private volatile double average = 0;
 
     public MovingAverage(int windowSize) {
-        if (windowSize <= 0) {
-            throw new IllegalArgumentException("window size must be greater than zero");
-        }
-
+        checkWindowSize(windowSize);
         this.windowSize = windowSize;
         this.observations = new long[windowSize];
     }
 
+    /**
+     * Used for changing the window size of {@code MovingAverage}.
+     *
+     * @param newWindowSize new window size.
+     * @return copy of original object with updated size.
+     */
+    public MovingAverage copyWithSize(int newWindowSize) {
+        MovingAverage copy = new MovingAverage(newWindowSize);
+        // Start is inclusive, but end is exclusive
+        long start, end = count;
+        if (isReady() == false) {
+            start = 0;
+        } else {
+            start = end - windowSize;
+        }
```

Comment:
won't  end be `(int) (count % observations.length - 1)`  ?

Reply:
end is exclusive (it's mentioned in the comment in line 38).

start and end at this time is some multiple k of the observations.length which is later handled in line 50.
```
for (int i = (int) start; i < end; i++) {
    copy.record(observations[i % observations.length]);
}
```